### PR TITLE
LPS-118146 Translate Custom Fields Data Type value

### PR DESCRIPTION
**Pull request by murilostodolni**
Related: https://issues.liferay.com/browse/LPS-118146

### CODE REVIEW

**Review by RafaelCRC — APPROVED** [review]
LGTM, since the Constants are hardcoded Strings i think its ok to get the language keys for this situation. what do you think @drewbrokke ?

**Comment by drewbrokke** [comment]
@RafaelCRC this fix will work, but we actually have a pattern for cases like this. See for example: https://github.com/liferay/liferay-portal/blob/c4821ede5dca7a6037b2198e9f81494dcc330f97/portal-kernel/src/com/liferay/portal/kernel/model/role/RoleConstants.java#L140-L155.

Basically, it's "give me the type, and I'll give you the relevant name".

Looks like ExpandoConstants has that for type, but not dataType:

https://github.com/liferay/liferay-portal/blob/9ffd9c6fe71c5ac3984173c5db007c042c62b339/portal-kernel/src/com/liferay/expando/kernel/model/ExpandoColumnConstants.java#L336-L402.

It would actually make sense to create a "getDataTypeLabel" method for this situation, since there are already dedicated keys for translating the different field types, and we can re-use some of those.

Thoughts?

File: modules/apps/expando/expando-web/src/main/resources/META-INF/resources/edit/default_value_input.jspf [hunk 1]
```diff
@@ -34,9 +34,9 @@
 	</c:when>
 	<c:when test="<%= propertyDisplayType.equals(ExpandoColumnConstants.PROPERTY_DISPLAY_TYPE_CHECKBOX) || propertyDisplayType.equals(ExpandoColumnConstants.PROPERTY_DISPLAY_TYPE_RADIO) || propertyDisplayType.equals(ExpandoColumnConstants.PROPERTY_DISPLAY_TYPE_SELECTION_LIST) %>">
 		<aui:select disabled="<%= expandoColumn != null %>" label="data-type" name="dataType" onChange="setDataType(true)" value="<%= ExpandoColumnConstants.getDataType(type) %>">
-			<aui:option label="<%= ExpandoColumnConstants.DATA_TYPE_TEXT %>" />
-			<aui:option label="<%= ExpandoColumnConstants.DATA_TYPE_DECIMAL %>" />
-			<aui:option label="<%= ExpandoColumnConstants.DATA_TYPE_INTEGER %>" />
+			<aui:option label="text" />
+			<aui:option label="decimal" />
+			<aui:option label="integer" />
```
Comment (drewbrokke) [inline modules/apps/expando/expando-web/src/main/resources/META-INF/resources/edit/default_value_input.jspf:39]: I think changing the label while no value is present will actually change the value, which could break this.

We need to include both a `label` and a `value`.

**Comment by RafaelCRC** [comment]
I agree, it would aggregate with expandoconstants following its patterns and its a better approach to work on a globally'ish solution. Can you do that @murilostodolni ?

**Comment by murilostodolni** [comment]
Hi @drewbrokke and @RafaelCRC 

In my opnion, the better is to add the atributte `value` in tag `aui:option`. In [this case](https://github.com/liferay/liferay-portal/blob/9ffd9c6fe71c5ac3984173c5db007c042c62b339/modules/apps/server/server-admin-web/src/main/resources/META-INF/resources/edit_document_library_extra_settings.jsp#:~:text=%3E%27%3E-,%3C%25,%25%3E,-%3C/aui%3Aselect%3E) the method make sense because there are too much types, but in this case there are three types only. If we create the method `getDataTypeLabel`, we need create the `DATA_TYPE_LABEL` array. What do you think @drewbrokke ? Add `value` or create the method `getDataTypeLabel`?

**Comment by drewbrokke** [comment]
@murilostodolni I think it's sensible to do both. We need to preserve the existing value in any case, so if we are going to adjust the label, at `value` attribute is required. And I think creating the method is also a good idea because it provides a single logical place to relate the data type to the label it uses. IMO it's fine to re-use one of the existing type labels found in the constants file, but it should also be ok to create our own new labels array.

**Comment by murilostodolni** [comment]
Hi @drewbrokke ok. I'm working in implementation now. Thank you.

**Comment by murilostodolni** [comment]
@drewbrokke one more think, i don't return the constants `DATA_TYPE_DECIMAL`, `DATA_TYPE_INTEGER`, `DATA_TYPE_TEXT` how they are now. Because the first letter is upper case and the language is lower case. In this case, if i modify, we can broke other parts the portal. What should i do?

**Comment by drewbrokke** [comment]
@murilostodolni you are correct, if you change the case, it will break the form submission. The `value` should still be the `DATA_TYPE_*` values, while the label should use the results of the new method.

**Comment by murilostodolni** [comment]
ci:test:sf

**Comment by murilostodolni** [comment]
ci:test:sf

**Comment by murilostodolni** [comment]
ci:test:relevant

**Comment by murilostodolni** [comment]
ci:test:relevant

**Comment by murilostodolni** [comment]
ci:test:relevant

**Comment by RafaelCRC** [comment]
@epark18 can you check if this failures are related? thanks in advance

File: portal-kernel/src/com/liferay/expando/kernel/model/ExpandoColumnConstants.java [hunk 2]
```diff
@@ -211,6 +217,20 @@ else if ((type == STRING) || (type == STRING_ARRAY) ||
...
+		}
+		else if (type == INTEGER) {
+			return DATA_TYPE_INTEGER;
+		}
+		else if (type == INDEX_TYPE_TEXT) {
+			return DATA_TYPE_TEXT;
+		}
+
+		return UNKNOWN_LABEL;
+	}
```
Comment (drewbrokke) [inline portal-kernel/src/com/liferay/expando/kernel/model/ExpandoColumnConstants.java:232]: This seems backwards to me. This method should accept one of the data types and return a label value. These are accepting a type and returning a data type.

**Review by drewbrokke — CHANGES_REQUESTED** [review]

File: portal-kernel/src/com/liferay/expando/kernel/model/ExpandoColumnConstants.java (hunk 2 above)
Comment (murilostodolni) [reply portal-kernel/src/com/liferay/expando/kernel/model/ExpandoColumnConstants.java:232]: Hi @drewbrokke. Sorry, but i don't understand. The method returnt [this values](https://github.com/liferay/liferay-portal/blob/master/portal-kernel/src/com/liferay/expando/kernel/model/ExpandoColumnConstants.java#:~:text=public%20static%20final%20String%20DATA_TYPE_DECIMAL,DATA_TYPE_TEXT%20%3D%20%22Text%22%3B). I created the method the same form the method [getTypeLabel](https://github.com/liferay/liferay-portal/blob/master/portal-kernel/src/com/liferay/expando/kernel/model/ExpandoColumnConstants.java#:~:text=%7D-,public%20static%20String%20getTypeLabel(int%20type)%20%7B,%7D,-private%20static%20final), the difference is that we need return the value "Text, Decimal and Integer" for the translation to work in tag `aui:option`. Sorry if i am understanding the wrong way. Thanks

Comment (drewbrokke) [reply portal-kernel/src/com/liferay/expando/kernel/model/ExpandoColumnConstants.java:232]: The `DATA_TYPE_*` fields are **data first**, and were only incidentally used as display labels. They are intended to be used as indicators of the data type a field should be. 

A label is strictly for display purposes and should not be used as any kind of data. 

So in this case, the `DATA_TYPE_*` values should populate the `value` attribute, while their respective display labels will be passed into the `label` attribute. 

This method should return what should be _displayed_, not what should be used as a value.

**Comment by drewbrokke** [comment]
@murilostodolni @RafaelCRC any updates on this one? Any blockers?

**Comment by murilostodolni** [comment]
Hi @drewbrokke i don't have new updates. I will closed this PR and will open a new when has update. Thanks

**Comment by murilostodolni** [comment]
ci:close

**Pull request closed without merging**